cmd/finch-daemon/main.go

```diff
@@ -104,32 +104,41 @@ func getListener(options *DaemonOptions) (net.Listener, error) {
 	return listener, nil
 }
 
-func handlePidFileOptions(options *DaemonOptions) error {
+func run(options *DaemonOptions) error {
+	// This sets the log level of the dependencies that use logrus (e.g., containerd library).
+	if options.debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
 	if options.pidFile != "" {
 		if err := os.MkdirAll(filepath.Dir(options.pidFile), 0o640); err != nil {
 			return fmt.Errorf("failed to create pidfile directory %s", err)
 		}
-		if err := pidfile.Write(options.pidFile, os.Getpid()); err != nil {
-			return fmt.Errorf("failed to start daemon, ensure finch daemon is not running or delete %s %w", options.pidFile, err)
+
+		pidFile, err := os.OpenFile(options.pidFile, os.O_RDWR|os.O_CREATE, 0o640)
+		if err != nil {
+			return fmt.Errorf("failed to open or create PID file: %w", err)
 		}
-	}
-	return nil
-}
 
-func run(options *DaemonOptions) error {
-	// This sets the log level of the dependencies that use logrus (e.g., containerd library).
-	if options.debug {
-		logrus.SetLevel(logrus.DebugLevel)
-	}
+		// Lock the PID file to prevent concurrent access
+		if err := syscall.Flock(int(pidFile.Fd()), syscall.LOCK_EX|syscall.LOCK_NB); err != nil {
```

Comment:
Can we use the [internal filelock pkg](https://pkg.go.dev/cmd/go/internal/lockedfile/internal/filelock) instead of syscall?

Reply:
it is part of the internal package, not sure it can be accessed and syscall.Flock seems to be in practice.

Reply:
We could use https://github.com/rogpeppe/go-internal (or similar), but I'm not sure if its worth it since this is only really supposed to run on Linux hosts. I'm good to close this comment pending action on https://github.com/golang/go/issues/33974 (seems like the proposal has been on hold for a long time now).

Reply:
ended up using https://github.com/gofrs/flock